flamingpy/noise/cv.py

```diff
@@ -0,0 +1,623 @@
+
+        # Assign state labels.
+        self.populate_states()
+
+    # Error correction methods
+    def apply_noise(self, sampling_order=None, rng=default_rng()):
+        """Apply cv-level noise to the graph state.
+
+        Identify the nodes of the EGraph with CV states, measure the syndrome
+        assuming the specified noise model, and convert the simulated homodyne
+        outcomes to bit values.
+
+        This method modifies the attributes of self.egraph.
+        """
+        if sampling_order is None:
+            sampling_order = self.sampling_order
+        self.measure_syndrome(sampling_order=sampling_order, rng=rng)
+        self.inner_decoder()
+
+    def populate_states(self, rng=default_rng()):
+        """Populate the graph state with state labels.
+
+        Assume the graph states consists of a combination of squeezed states
+        and GKP+ states. A non-zero self.p_swap overrides indices specified in
+        self.states and uses a binomial distribution to identify some indices
+        as p-squeezed states.
+
+        Raise an exception if both p_swap and p indices are supplied.
+
+        This method modifies self.egraph.
+        """
+        # Generate indices of squeezed states based on swap-out
+        # probability self.p_swap, if supplied.
+        if self.p_swap:
+            self._generate_squeezed_indices(rng)
+
+        # Associate remaining indices with GKP states.
+        self._generate_gkp_indices()
+
+        # Give the EGraph nodes state attributes.
+        self._apply_state_labels()
+
+    def measure_syndrome(self, sampling_order="initial", rng=default_rng()):
+        """Measure the syndrome for memory-mode error correction of the grpah
+        states."""
+        return self.measure_hom(
+            quad="p", inds=self.code.all_syndrome_inds, sampling_order=sampling_order, rng=rng
+        )
+
+    def inner_decoder(self, translator=GKP_binner):
+        """Convert homodyne outcomes to bit values according to translator.
+
+        This is the inner (CV) decoder, a.k.a. translator, a.k.a binning function.
+        Set converted values to the bit_val attribute for nodes in self.egraph.
+
+        Args:
+            code (SurfaceCode): the qubit QEC code
+            translator (func): the choice of binning function; by default, the
+                standard GKP binning function that snaps to the closest
+                integer multiple of sqrt(pi).
+        Returns:
+            None
+        """
+        for point in self.code.all_syndrome_coords:
+            hom_val = self.code.graph.nodes[point]["hom_val_p"]
+            bit_val = translator([hom_val])[0]
+            self.code.graph.nodes[point]["bit_val"] = bit_val
+
+    # Public CV manipulation methods.
+    def measure_hom(
+        self,
+        quad="p",
+        inds=None,
+        updated_means=None,
+        updated_covs=None,
+        propagate=True,
+        sampling_order="initial",
+        rng=default_rng(),
+    ):
```

Comment:
This has so many parameters. It should become its own class. Or be divided in smaller methods.

Reply:
We will likely address this in a subsequent PR.